### ISSUE TITLE
How can I route way in both direction? 

### ISSUE THREAD

**Issue by Guo-astro** [valhalla/valhalla#1802]
The same question as in https://github.com/Project-OSRM/osrm-backend/issues/5438

In our case, I want to calculate  4 routes as in the picture(start heading in 2 directions and target heading in 2 directions)
 
![OSRMIssue 001](https://user-images.githubusercontent.com/15899028/56854976-26d60e80-697a-11e9-8f94-f3394fa31717.jpeg)

I guess the solution here is also modifying the insertion logic of heap in  Bi A*? 
Any hint will be helpful!

**Comment by kevinkreiser** [comment]
Right now, we already include all 4 edges in the algorithm and the algorithm can find all the paths, the issue is that we return what is the shortest one and we may not let the search continue searching before we say we've found the shortest one.

My question would be, do you want all 4 paths? Do you actually need them or do you want to in the end select just 1? I could see wanting all 4 options if you need to optimize a large list of s,t pairs so that you can make sure the way you come into a given stop is the way you exit it (avoid uturns).

**Comment by Guo-astro** [comment]
@kevinkreiser Thanks very much for the response!
I want all the 4 paths. The use case is exactly to avoid u-turns at stops when solving vrp problems.
I need to obtain the distance/duration matrix with all heading of the 4 paths.
The best solution I can think is to first implement it using the route service because it is easier to modify the api response. And next, implement it in the matrix service.
Any hint for the position of the line of code will be helpful!

**Comment by kevinkreiser** [comment]
@Guo-astro you can get all 4 routes with very few changes to the API by making a few api calls.

1. Modify `loki/search.cc` to add the heading to the candidate location, and make sure to serialize this information in the `locate` api
2. Call the `/locate` api for your input locations. Make sure to call it with `"verbose": true`. This will give you the candidate edges at both locations including their heading. Please note that you may get more than 2 edge candidates for any given location so you'll want to settle on just 2 per location. Additionally, you may only get one edge candidate for a given location because the street might be a one-way.
3. Make up to 4 calls to the `/sources_to_targets` api passing in all the combinations of headings for the start and end locations

With respect to your questions:

0. I would not modify this structure. What you want to do modify your request so that each heading is a different source/target. So if you have 2 locations A and B, you want to hit the `sources_to_targets` api with 2 sources and 2 targets, namely `sources = A.heading, A.180_heading` and `targets = B.heading, B.180_heading`.
1. edge ids are internal ids and do not match the osm way ids. we do keep this information though and you can access it but its stored in a way that is not performant to access during routing computations.
2. As described above I would separate the locations by heading so that you get individual results for all combinations.

**Comment by Guo-astro** [comment]
@kevinkreiser  Thanks very much for the answers!
1.Now I read into the code [here](https://github.com/valhalla/valhalla/blob/74f39e97f03dc6e540318586e3bba9cd47a49664/src/midgard/util.cc#L91)     What is the idea behind the code? Any explanation or docs will be helpful
2. In the [GetOffsetForHeading](https://github.com/valhalla/valhalla/blob/74f39e97f03dc6e540318586e3bba9cd47a49664/valhalla/baldr/graphconstants.h#L490 ) I can see some coefficients for different road class,are the value of those coefficients by experience?
3. Should I modify the [PathLocation](https://github.com/valhalla/valhalla/blob/74f39e97f03dc6e540318586e3bba9cd47a49664/valhalla/baldr/pathlocation.h#L22) structure with the heading variable ?    And then calculate the heading inside function [correlate_node& correlate_edge](https://github.com/valhalla/valhalla/blob/74f39e97f03dc6e540318586e3bba9cd47a49664/src/loki/search.cc#L726)?   Is it the best way to do it?

Thanks for the time!

**Comment by kevinkreiser** [comment]
@Guo-astro so the modification i meant was to change this function:https://github.com/valhalla/valhalla/blob/master/src/loki/search.cc#L46-L68

so that it actually takes the measurement of the heading it gets: https://github.com/valhalla/valhalla/blob/master/src/loki/search.cc#L57

and stores it on the PathEdge. to do that you'll need to add a heading to the path edges data structure (both the protobuf https://github.com/valhalla/valhalla/blob/master/proto/tripcommon.proto#L35 and the c++ class https://github.com/valhalla/valhalla/blob/master/valhalla/baldr/pathlocation.h#L31)

after you've made sure that loki::Search always fills out the heading on the path edge, you'll want to make sure that the locate service serializes this information out. to do that head over to here: https://github.com/valhalla/valhalla/blob/master/src/tyr/locate_serializer.cc#L31

now you should have yourself all the information you need coming back in the locate api. such that you can make the matrix request with all the permutations you need.

also please note that the locate service is a batch service and benefits from having a high number of coordinates per call (if you are doing a large number of requests). so one thing you can do is send 100s of locations to locate up front instead of doing only 2 locations per call to locate.

**Comment by Guo-astro** [comment]
Thanks again for the reply! 
I have done the loki part and can obtain all the heading on a snapped point.
Now I want to modify the /sources_to_targets request api.
As previously mentioned, "So if you have 2 locations A and B, you want to hit the sources_to_targets api with 2 sources and 2 targets, namely sources = A.heading, A.180_heading and targets = B.heading, B.180_heading."   I need to 
1. Modify the request to include headings [here](https://github.com/valhalla/valhalla/blob/e01cf6e2cd8cb78106f70e590bf20a7cb7dfc347/src/thor/worker.cc#L110)
2.When the request comes into the SourceToTarget function, I need to modify [SetOriginOneToMany](https://github.com/valhalla/valhalla/blob/e01cf6e2cd8cb78106f70e590bf20a7cb7dfc347/src/thor/timedistancematrix.cc#L408)  [SetOriginManyToOne](https://github.com/valhalla/valhalla/blob/e01cf6e2cd8cb78106f70e590bf20a7cb7dfc347/src/thor/timedistancematrix.cc#L462)  [SetDestinations](https://github.com/valhalla/valhalla/blob/e01cf6e2cd8cb78106f70e590bf20a7cb7dfc347/src/thor/timedistancematrix.cc#L512) [SetDestinationsManyToOne](https://github.com/valhalla/valhalla/blob/e01cf6e2cd8cb78106f70e590bf20a7cb7dfc347/src/thor/timedistancematrix.cc#L561) to use heading information  from edgePath. any suggestion?

**Comment by kevinkreiser** [comment]
No further modifications should be needed. After the first modification where you add the heading to the path edges you should be able to:

1. call localhost:8002/locate to get the different headings for a given location
2. call localhost:8002/sources_to_targes with the different combinations of headings where each heading is its own location.

**Comment by Guo-astro** [comment]
Thanks very much!
I still didn’t  get the 2nd part.
For example I have the following request:

`curl http://localhost:8002/sources_to_targets --data '{"sources":[{"lat":35.169715,"lon":136.931980}],"targets":[{"lat":35.169419,"lon":136.931899},{"lat":35.169419,"lon":136.931899}],"costing":"auto"}'`

How should I modify it? Do you mean I can pass heading in the request?

**Comment by kevinkreiser** [comment]
right so this is what you do:

    curl http://localhost:8002/locate --data '{"locations":[{"lat":35.169715,"lon":136.931980},{"lat":35.169419,"lon":136.931899},{"lat":35.169419,"lon":136.931899}],"costing":"auto","verbose":true}'

from there you will have the results, of the snapping and because of your modifications the headings with each one. now you can do:

    curl http://localhost:8002/sources_to_targets --data '{"sources":[{"lat":35.169715,"lon":136.931980,"heading":123},{"lat":35.169715,"lon":136.931980,"heading":303}],"targets":[{"lat":35.169419,"lon":136.931899,"heading":25},{"lat":35.169419,"lon":136.931899,"heading":205}],"costing":"auto"}'

which will give all 4 permutations

**Comment by Guo-astro** [comment]
Works like a charm!
1. I would like to open a pr so you can check the code !
2. In matrix docs I didn’t find we can use the heading paramaters is it under construction? I would like to help with filling in the docs too!
Thank you!

**Comment by kevinkreiser** [comment]
The location parameters work for all end points, in think there is a line in the docs that says that... but I have to check!

**Comment by Guo-astro** [comment]
Will it be good if using a swagger file for all the apis?

**Comment by kevinkreiser** [comment]
I have no experience with swagger unfortunately so I wont be able to advise you on that one, sorry about that!

**Comment by Guo-astro** [comment]
@kevinkreiser   I tried use the real data and heading with
`{"sources":[{"lon":136.924255,"lat":35.186695,"heading":0},{"lon":136.924255,"lat":35.186695,"heading":180}],"targets":[{"lon":136.923218,"lat":35.186218,"heading":270},{"lon":136.923218,"lat":35.186218,"heading":90}],"costing":"auto"}`

The response I got `{
    "sources": [
        [
            {
                "lat": 35.186695, 
                "lon": 136.924255
            }, 
            {
                "lat": 35.186695, 
                "lon": 136.924255
            }
        ]
    ], 
    "sources_to_targets": [
        [
            {
                "distance": 0.148, 
                "from_index": 0, 
                "time": 41, 
                "to_index": 0
            }, 
            {
                "distance": 0.148, 
                "from_index": 0, 
                "time": 41, 
                "to_index": 1
            }
        ], 
        [
            {
                "distance": 0.148, 
                "from_index": 1, 
                "time": 41, 
                "to_index": 0
            }, 
            {
                "distance": 0.148, 
                "from_index": 1, 
                "time": 41, 
                "to_index": 1
            }
        ]
    ], 
    "targets": [
        [
            {
                "lat": 35.186218, 
                "lon": 136.923218
            }, 
            {
                "lat": 35.186218, 
                "lon": 136.923218
            }
        ]
    ], 
    "units": "kilometers"
}
`
All the distances are the same, any idea?

**Comment by kevinkreiser** [comment]
Yep, please remove that and the other one on the targets a couple lines further down. Thanks!

**Comment by Guo-astro** [comment]
By the way, what is the difference between costmatrix and timedistancematrix?

**Comment by dnesbitt61** [comment]
TmeDistanceMatrix is optimized for one to many or many to one matrix calls. To compute many to many it basically makes several calls to one to many, which is somewhat inefficient. CostMatrix is optimized to compute many to many matrices but it takes more memory and is slower for some types of problems (when there are many locations close to each other for pedestrian and bicycling).

**Comment by Guo-astro** [comment]
@dnesbitt61 Thanks very much for the reply!
I ran some test and found out the CostMatrix is performance a little worse than TmeDistanceMatrix.
Both figures are using the same data. The vertical axis is calculation time in seconds, the horizontal axis is the number of coords,i.e., 300 means 300*300 elements. The mode is "auto". 
![image](https://user-images.githubusercontent.com/15899028/58060500-98d1fa00-7bac-11e9-9a29-06abf3e7a8e1.png)
From the code of CostMatrix, I guess it is using a bidirectional search approach?

**Comment by dnesbitt61** [comment]
yes, CostMatrix uses bidirectional search. I found CostMatrix performed better for driving routes when the locations were further apart. Both methods take a long time when locations are very far apart.

**Comment by nilsnolde** [comment]
Good question, seems weird and no idea off the top of my head. Maybe @dnesbitt61? 

Until we know, you can try to remove it, run the tests and/or some matrices before/after.

**Comment by nilsnolde** [comment]
Just raise a PR with that, the tests will tell us some. Still can't think of any reason why we'd clear headings..

**Comment by jehamannco** [comment]
> Just raise a PR with that, the tests will tell us some. Still can't think of any reason why we'd clear headings..

hi, i tried to push a branch mit the change in order to raise the PR but i get the following message:

remote: Permission to valhalla/valhalla.git denied to jehamannco.

**Comment by nilsnolde** [comment]
You'll need to fork the repository, create a branch there, edit the source files, push the changes to your fork and then raise a PR. There's Github docs how to contribute to a project. Otherwise we'll do it real quick, it's minimal.
